File: c_src/fswatcher.c
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

// This is not really used...

#include <stdio.h>
#include <limits.h>
#include <stdlib.h>
#include <errno.h>
#include <sys/types.h>
#include <linux/inotify.h>

#define EVENT_SIZE  ( sizeof (struct inotify_event) )
#define EVENT_BUF_LEN     (128 * (EVENT_SIZE+16+NAME_MAX))

int main(int argc, char *argv[])
{
  if (argc < 2)
  {
    printf("Usage: %s root/path/to/watch /anotherpath/to/watch  .. \r\n",argv[0]);
    return EXIT_SUCCESS;
  }
  int length, i = 0;
  int notifyfd;
  int wd;
  char buffer[EVENT_BUF_LEN];
  int watches[argc-1];
  fd_set selectset;
  FD_ZERO(&selectset);
  FD_SET(0, &selectset);

  /*creating the INOTIFY instance*/
  notifyfd = inotify_init();
  FD_SET(notifyfd, &selectset);

  /*checking for error*/
  if ( notifyfd < 0 ) 
  {
    perror("inotify_init failed");
  }

  for (i = 1; i < argc; i++)
  {
    watches[i-1] = inotify_add_watch(notifyfd, argv[i], IN_CREATE | IN_MODIFY);
  }
    

  
  for (;;)
  {
    fd_set readset = selectset;
    if (select(notifyfd+1, &readset, NULL, NULL, NULL) == -1)
    {
      perror("Unable to select");
      return 0;
    }
    
    if (FD_ISSET(0,&readset))
      break;
    else
    {
      length = read(notifyfd, buffer, EVENT_BUF_LEN); 
      if (length < 0) 
      {
        perror("unable to read from inotify");
        return 0;
      }  

      i = 0;
      while(i < length) 
      {     
        struct inotify_event *event = (struct inotify_event*)&buffer[i];     
        if (event->len)
        {
          int j;
          for (j = 1; j < argc; j++)
          {
            if (watches[j-1] == event->wd)
            {
              printf("%s/%s\r\n", argv[j], event->name);
              break;
            }
          }
        }
        i += EVENT_SIZE + event->len;
      }
    }
  }

  for (i = 1; i < argc; i++)
    inotify_rm_watch(notifyfd, watches[i-1]);
  close(notifyfd);
}

File: c_src/fswatcher.m
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

// Console app that prints path to folder where a file has changed.
#include <objc/Object.h>
#include <stdio.h>
#include <stdlib.h>
#include <CoreFoundation/CFArray.h>
#include <Foundation/NSArray.h>
#include <Foundation/NSString.h>
#include <CoreFoundation/CFRunLoop.h>
#include <CoreFoundation/CFUUID.h>
#include <CoreServices/CoreServices.h>
#include <CoreFoundation/CFDate.h>

// gcc -lobjc -framework CoreFoundation -framework CoreServices c_src/fswatcher.m -o /priv/fswatcher

void EventStreamCallback(ConstFSEventStreamRef sref, void* user_data, size_t numEvents, void* eventPaths,
                                  const FSEventStreamEventFlags sflags[], const FSEventStreamEventId seid[]);

void EventStreamCallback(ConstFSEventStreamRef sref, void* user_data, size_t numEvents, void* eventPaths,
                                  const FSEventStreamEventFlags sflags[], const FSEventStreamEventId seid[])
{
    char** paths = (char**)(eventPaths);

    int i = 0;
    for (i = 0; i < numEvents; ++i)
    {
        printf("%s\r\n",paths[i]);
    }
    fflush(stdout);
}
static void user_input(CFSocketRef s, CFSocketCallBackType type, 
	   CFDataRef address, const void *data, void *info)
{
	char str[100];
	char* ret = fgets(str, sizeof(str), stdin);
	if (ret == NULL || !CFSocketIsValid(s))
	{
		CFRunLoopStop(CFRunLoopGetCurrent());
	}
}





int main(int argc, char *argv[])
{
  int i = 0;
	if (argc <= 1)
	{
		printf("Usage: %s root/path/to/watch\r\n",argv[0]);
		return EXIT_SUCCESS;
	}

  for (i = 1; i < argc; i++)
  {
    CFAbsoluteTime latency = 0.5;
    CFStringRef pathstr  = CFStringCreateWithCString(kCFAllocatorDefault, argv[i], kCFStringEncodingUTF8);
    CFTypeRef   cfValues[] = { pathstr };
    CFArrayRef paths = CFArrayCreate( kCFAllocatorDefault, cfValues, 1, &kCFTypeArrayCallBacks );

    FSEventStreamRef fsstream = FSEventStreamCreate(NULL,
           &EventStreamCallback,
           NULL,
           paths,
           kFSEventStreamEventIdSinceNow,
           latency,
           kFSEventStreamCreateFlagNone
     );

    FSEventStreamScheduleWithRunLoop(fsstream, CFRunLoopGetCurrent(), kCFRunLoopDefaultMode);
    FSEventStreamStart(fsstream);
  }

  CFSocketRef	socket;
  socket = CFSocketCreateWithNative(NULL, fileno(stdin), kCFSocketReadCallBack, user_input, NULL);
  CFRunLoopSourceRef rls = CFSocketCreateRunLoopSource(NULL, socket, 0);
  CFRunLoopAddSource(CFRunLoopGetCurrent(), rls, kCFRunLoopDefaultMode);

  CFRunLoopRun();

	return EXIT_SUCCESS;
}




